### PR TITLE
move autofocus process to postFrame phase

### LINKED ISSUE #84845: `FocusNode.autofocus` should not skip when the previously focused child in the focus scope is about to unmount.
### Repro

```dart
testWidgets('autofocus', (WidgetTester tester) async {
  final Widget autofocus = Focus(autofocus: true, child: const Placeholder());

  await tester.pumpWidget(FocusScope(child: autofocus));
  debugDumpFocusTree();
  await tester.pumpWidget(FocusScope(child: SizedBox(child: autofocus)));
  debugDumpFocusTree();
});
```
#### initial focus tree: 
```
FocusManager#40da4
 │ primaryFocus: FocusNode#c9633([PRIMARY FOCUS])
 │ primaryFocusCreator: Focus ← _FocusMarker ← Semantics ←
 │   FocusScope ← [root]
 │
 └─rootScope: FocusScopeNode#e2ba0(Root Focus Scope [IN FOCUS PATH])
   │ IN FOCUS PATH
   │ focusedChildren: FocusScopeNode#8b87b([IN FOCUS PATH])
   │
   └─Child 1: FocusScopeNode#8b87b([IN FOCUS PATH])
     │ context: FocusScope
     │ IN FOCUS PATH
     │ focusedChildren: FocusNode#c9633([PRIMARY FOCUS])
     │
     └─Child 1: FocusNode#c9633([PRIMARY FOCUS])
         context: Focus
         PRIMARY FOCUS
```

#### updated focus tree 

```
FocusManager#40da4
 │ primaryFocus: FocusScopeNode#8b87b([PRIMARY FOCUS])
 │ primaryFocusCreator: FocusScope ← [root]
 │
 └─rootScope: FocusScopeNode#e2ba0(Root Focus Scope [IN FOCUS PATH])
   │ IN FOCUS PATH
   │ focusedChildren: FocusScopeNode#8b87b([PRIMARY FOCUS])
   │
   └─Child 1: FocusScopeNode#8b87b([PRIMARY FOCUS])
     │ context: FocusScope
     │ PRIMARY FOCUS
     │
     └─Child 1: FocusNode#9f1d7
         context: Focus
```

I'd expect `FocusNode#9f1d7` to get the focus because the previously focused widget is now gone:
>   /// True if this widget will be selected as the initial focus when no other
  /// node in its scope is currently focused.
  ///
  /// Ideally, there is only one widget with autofocus set in each [FocusScope].
  /// If there is more than one widget with autofocus set, then the first one
  /// added to the tree will get focus.

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes #84845 

There is a similar logic in  https://github.com/flutter/flutter/blob/64401fb8dfce47bb0b31f9d513adb10a61c5451c/packages/flutter/lib/src/widgets/editable_text.dart#L1610

### CODE REVIEW

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter/lib/src/widgets/focus_scope.dart [hunk 1]
```diff
@@ -614,8 +615,13 @@ class _FocusState extends State<Focus> {
 
   void _handleAutofocus() {
     if (!_didAutofocus && widget.autofocus) {
-      FocusScope.of(context).autofocus(focusNode);
       _didAutofocus = true;
+      // Waiting for inactive nodes to be disposed.
+      SchedulerBinding.instance!.addPostFrameCallback((_) {
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/focus_scope.dart:620]: I think ideally we should fix `FocusScopeNode.autofocus`.

File: packages/flutter/lib/src/widgets/focus_scope.dart [hunk 2]
```diff
@@ -614,8 +615,13 @@ class _FocusState extends State<Focus> {
...
+        if (mounted) {
+          FocusScope.of(context).autofocus(focusNode);
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/focus_scope.dart:622]: Is it possible for this `focusNode` to be disposed?

Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/focus_scope.dart:622]: It's probably not possible, since the widget hosting the node is mounted.

Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/focus_scope.dart:622]: Oh sorry I thought this line was in `FocusScopeState`. Although very unlikely but it still can happen if someone schedules a `focsusNode.dispose` in a postframe callback right?

File: packages/flutter/lib/src/widgets/focus_scope.dart (hunk 1 above)
Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/focus_scope.dart:620]: I tried it, but the inactive node may become active again.
 For example, by the global key or the node's attachment is rewritten, so we need to wait until finalizetree to confirm whether the pending node has a chance to regain focus, which is not very easy to do.

File: packages/flutter/lib/src/widgets/focus_scope.dart (hunk 2 above)
Comment (xu-baolin) [reply packages/flutter/lib/src/widgets/focus_scope.dart:622]: I think the life cycle of `focusNode` and `_FcusState` should be consistent

**Comment by goderbauer** [comment]
(PR triage) @LongCatIsLooong @gspencergoog @xu-baolin Are there still plans to continue with this PR?

**Comment by xu-baolin** [comment]
Continue work via #85562

**Pull request closed without merging**